cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/BasicRunnableContext.java

```diff
@@ -0,0 +1,250 @@
+  private final StreamWriter streamWriter;
+
+  public BasicRunnableContext(T spec, Set<String> datasets, Program program, RunId runId, int instanceId,
+                              Arguments runtimeArgs, CConfiguration cConf,
+                              MetricsCollectionService metricsCollectionService,
+                              DatasetFramework datasetFramework,
+                              TransactionSystemClient transactionSystemClient,
+                              DiscoveryServiceClient discoveryServiceClient,
+                              StreamWriterFactory streamWriterFactory,
+                              @Nullable AdapterDefinition adapterSpec,
+                              @Nullable PluginInstantiator pluginInstantiator) {
+    super(program, runId, runtimeArgs, datasets,
+          getMetricCollector(program, runId.getId(), instanceId, metricsCollectionService, adapterSpec),
+          datasetFramework, discoveryServiceClient, adapterSpec, pluginInstantiator);
+    this.program = program;
+    this.instanceId = instanceId;
+    this.transactionSystemClient = transactionSystemClient;
+    this.datasetFramework = datasetFramework;
+    this.loggingContext = createLoggingContext(program.getId(), runId, adapterSpec);
+    if (metricsCollectionService != null) {
+      this.userMetrics = new ProgramUserMetrics(getProgramMetrics());
+    } else {
+      this.userMetrics = null;
+    }
+    this.runtimeArgs = runtimeArgs.asMap();
+    this.streamWriter = streamWriterFactory.create(program.getId().getNamespace(), getOwners());
+
+    // The cache expiry should be greater than (2 * transaction.timeout) and at least 2 hours.
+    // This ensures that when a dataset instance is requested multiple times during a single transaction,
+    // the same instance is always returned.
+    long cacheExpiryTimeout =
+      Math.max(2, 2 * TimeUnit.SECONDS.toHours(cConf.getInt(TxConstants.Manager.CFG_TX_TIMEOUT,
+                                                            TxConstants.Manager.DEFAULT_TX_TIMEOUT)));
+    // A cache of datasets by threadId. Repeated requests for a dataset from the same thread returns the same
+    // instance, thus avoiding the overhead of creating a new instance for every request.
+    this.datasetsCache = CacheBuilder.newBuilder()
+      .expireAfterAccess(cacheExpiryTimeout, TimeUnit.HOURS)
+      .removalListener(new RemovalListener<Long, Map<DatasetCacheKey, Dataset>>() {
+        @Override
+        @ParametersAreNonnullByDefault
+        public void onRemoval(RemovalNotification<Long, Map<DatasetCacheKey, Dataset>> notification) {
+          if (notification.getValue() != null) {
+            for (Map.Entry<DatasetCacheKey, Dataset> entry : notification.getValue().entrySet()) {
+              try {
+                entry.getValue().close();
+              } catch (IOException e) {
+                LOG.error("Error closing dataset: {}", entry.getKey(), e);
```

Comment:
What's this mean?

I didn't write this code, because this was directly copied over from the original worker implementation